src/core/dappletEngine.ts

```diff
@@ -1,18 +1,14 @@
 import { DappletFrameExecutor } from "./dappletFrameExecutor";
 import { DappletRequest } from "../types/dappletRequest";
-import { DappletProvider } from "../interfaces/dappletProvider";
-import { FrameStatus } from "../types/statusEnum";
-import { FeatureRegistry } from './featureRegistry';
 import { DappletContext } from './dappletContext';
 
 export class DappletEngine {
     private _frameExecutors: DappletFrameExecutor[] = [];
-    private _statusChangedHandler: (statuses: FrameStatus[]) => void = () => {};
 
     constructor(dappletRequest: DappletRequest, private _context: DappletContext) {
         for (const frame of dappletRequest.frames) {
-            const executor = new DappletFrameExecutor(frame.dappletId, frame.txMeta, this._context.featureRegistry);
-            executor.onStatusChanged((s) => this._newExecutorStatus(executor, s));
+            if (!frame.dapplet) throw Error("Dapplet is not loaded.");
+            const executor = new DappletFrameExecutor(frame.dapplet, frame.txMeta, this._context.featureRegistry);
             this._frameExecutors.push(executor);
         }
     }
```

Comment:
as I mentioned earlier, I wouldn't re-use request as an input parameter, so you don't need to check if Dapplet is not loaded. Just try to pass an object you need instead of checking the container object.